hazelcast/src/main/java/com/hazelcast/nio/serialization/ClassDefinitionBuilder.java

```diff
@@ -213,7 +341,10 @@ public ClassDefinition build() {
         return cd;
     }
 
-    private void check() {
+    private void check(String fieldName) {
+        if (!addedFieldNames.add(fieldName)) {
```

Comment:
Forgot to mention one potential problem with this approach. In situation when `#addPortableArrayField()` or `#addPortableField()` are invoked with incorrect ClassDefinition definition (with zero classId), the field will be added into the `addedFieldNames`, while, in fact, it wasn't added due to next failed check. The simplest fix would be to remove the field before throwing `IllegalArgumentException` in those methods, but there might be more elegant solutions.

Reply:
Good catch. I fixed it.

Reply:
I'd say it doesn't matter. As soon as the builder throws an exception, it should be considered broken. A code that catches an exception and continue with the builder is very rare I'd say...

Reply:
It's normal to keep using the builder after a validation exception was thrown. If we consider the object broken, we must enforce this contract and throw on any subsequent call. A public API shouldn't be based on fuzzy assumptions which only exist in maintainers minds.